lib/v2/screens/sign_up/create_username/create_username.dart

```diff
@@ -82,18 +99,40 @@ class _CreateUsernameState extends State<CreateUsername> {
     );
   }
 
+  String? _validateUsername(String? username) {
+    final validCharacters = RegExp(r'^[a-z1-5]+$');
+
+    if (username.isNullOrEmpty) {
+      return 'Please select a username';
+    } else if (RegExp(r'0|6|7|8|9').allMatches(username!).isNotEmpty) {
+      return 'Only numbers 1-5 can be used.';
+    } else if (username.toLowerCase() != username) {
+      return "Name can be lowercase only";
+    } else if (!validCharacters.hasMatch(username) || username.contains(' ')) {
+      return "No special characters or spaces can be used";
+    } else if (username.length != 12) {
+      return 'Username must be 12 characters long';
+    }
+
+    return null;
+  }
```

Comment:
This should not be in the UI. How about creating a utility class for strigs with validation?

```
class StringField {
  final String value;
  final bool isRequired;
  final Pattern? pattern;
  bool? isValid;
  String? error;

  StringField({this.value = ''})..validate();

  void validate() {
    isValid = true;
    error = null;

    if (isRequired) {
      if (value.trim().length == 0) {
        isValid = false;
        error = requiredErrorMessage;
      }
    }

    if (value.isEmpty == false && pattern != null && isValid!) {
      RegExp regex = RegExp(pattern as String);

      isValid = regex.hasMatch(value);
      if (!isValid!) {
        error = patternErrorMessage;
      }
    }

    if (value.isEmpty == false && isValid! && minLengthRequired != -1) {
      if (value.length < minLengthRequired) {
        isValid = false;
        error = minLengthErrorMessage;
      }
    }
    }
  }

  String getValue() => originalValue != value.trim() ? value : '';

  StringField copyWith({String? value}) {
    return StringField(
      value: value ?? this.value,
      isRequired: isRequired,
      pattern: pattern,
    );
  }

  @override
  String toString() => 'StringField value: $value';
}
```

```
//BlocState
final StringField username;
```

```
// Bloc
yield state.copyWith(userame: state.username.copyWith(value: event.username));
```

```
// UI
validator: state.username.error
```

Reply:
I think it will be overkill and also the string validation here is not generic. I moved it into the bloc anyways, so it's not in the UI.

Reply:
agree with farzad on this one